sugar.go

```diff
@@ -0,0 +1,214 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+import (
+	"fmt"
+
+	"go.uber.org/zap/zapcore"
+)
+
+const oddNumberErrMsg = "Passed an odd number of keys and values to SugaredLogger, ignoring last."
+
+// A SugaredLogger wraps the core Logger functionality in a slower, but less
+// verbose, API.
+type SugaredLogger struct {
+	core Logger
+}
+
+// Sugar converts a Logger to a SugaredLogger.
+func Sugar(core Logger) *SugaredLogger {
+	// TODO: increment caller skip.
+	return &SugaredLogger{core}
+}
+
+// Desugar unwraps a SugaredLogger, exposing the original Logger.
+func Desugar(s *SugaredLogger) Logger {
+	// TODO: decrement caller skip.
+	return s.core
+}
+
+// With adds a variadic number of key-value pairs to the logging context.
+// Even-indexed arguments are treated as keys, and are converted to strings
+// (with fmt.Sprint) if necessary. The keys are then zipped with the
+// odd-indexed values into typed fields using the Any field constructor.
+//
+// For example,
+//   sugaredLogger.With(
+//     "hello", "world",
+//     "failure", errors.New("oh no"),
+//     "count", 42,
+//     "user", User{name: "alice"},
+//  )
+// is the equivalent of
+//   coreLogger.With(
+//     String("hello", "world"),
+//     String("failure", "oh no"),
+//     Int("count", 42),
+//     Object("user", User{name: "alice"}),
+//   )
+func (s *SugaredLogger) With(args ...interface{}) *SugaredLogger {
+	return s.WithFields(sweetenFields(args, s.core)...)
+}
+
+// WithFields adds structured fields to the logger's context, just like the
+// standard Logger's With method.
+func (s *SugaredLogger) WithFields(fs ...zapcore.Field) *SugaredLogger {
+	return &SugaredLogger{core: s.core.With(fs...)}
+}
+
+// Debug logs a message and some key-value pairs at DebugLevel. Keys and values
+// are treated as they are in the With method.
+func (s *SugaredLogger) Debug(msg interface{}, keysAndValues ...interface{}) {
+	if ce := s.core.Check(DebugLevel, sweetenMsg(msg)); ce != nil {
```

Comment:
The caller skipping should allow skipping n frames, not just one. It seems a little arbitrary to say a logger can only skip a single call frame.